src/main/kotlin/com/parmet/buf/gradle/GenerateTask.kt

```diff
@@ -26,10 +26,43 @@ abstract class GenerateTask : DefaultTask() {
         execBuf(args + additionalArgs())
     }
 
-    private fun additionalArgs() =
-        if (getExtension().generateOptions?.includeImports == true) {
+    private fun additionalArgs(): List<String> {
+        val generateOptions = getExtension().generateOptions
+        val importOptions = if (generateOptions?.includeImports == true) {
             listOf("--include-imports")
         } else {
             emptyList()
         }
+
+        val configOptions = genConfigFile()?.let {
+            listOf("--template", it.absolutePath)
+        } ?: emptyList()
+
+        return importOptions + configOptions
+    }
+
+    private fun genConfigFile(): File? {
+        return getExtension().generateOptions?.let { generateOptions ->
+            val selectedTemplateFiles = listOfNotNull(generateOptions.templateFileLocation, project.file("buf.gen.yaml")).filter {
+                it.exists() && it.isFile
+            }
+            check(selectedTemplateFiles.isNotEmpty()) {
+                if (generateOptions.templateFileLocation != null) {
+                    "Specified templateFileLocation does not exist."
+                } else {
+                    "No buf.gen.yaml file found in the root directory."
+                }
+            }
+            check(selectedTemplateFiles.size == 1) {
+                "Buf gen template file specified in the root directory as well as with templateFileLocation; pick one."
+            }
```

Comment:
I think this may look better condensed to a single tree, but I'm having trouble finding a branch structure I really like. I'm imagining something that looks like the test cases:

- if file is specified
    - the default file should not exist
    - the specified file had better exist and be a file
- otherwise the default file must exist and be a file

Reply:
If you agree, I will go ahead and add a helper extension to make the code read cleanly. The resultant method would look like this:

```kotlin
private fun resolveTemplateFile(): File? {
        return getExtension().generateOptions?.let { generateOptions ->
            val defaultTemplateFile = project.file("buf.gen.yaml").validOrNull()
            if (generateOptions.templateFileLocation != null) {
                val specifiedTemplateFile = generateOptions.templateFileLocation.validOrNull()
                check(specifiedTemplateFile != null) {
                    "Specified templateFileLocation does not exist."
                }
                check(defaultTemplateFile == null) {
                    "Buf gen template file specified in the root directory as well as with templateFileLocation; pick one."
                }
                specifiedTemplateFile
            } else {
                check(defaultTemplateFile != null) {
                    "No buf.gen.yaml file found in the root directory."
                }
                defaultTemplateFile
            }
        }
    }

    private fun File?.validOrNull() =
        if (this != null && isFile && exists())
            this
        else
            null
```

Reply:
I like that structure. Only changes I would make:

- The error message should probably be `"No buf.gen.yaml file found in the project directory."` instead of `root directory` since we're looking relative to the current project where the plugin has been applied
- `validOrNull` can be succinctly written in terms of `takeIf`: 
    ```kotlin
    private fun File?.validOrNull() =
        this?.takeIf { it.isFile && it.exists()) }
    ```